drivers/modem/modem_socket.c

```
@@ -342,6 +342,8 @@ int modem_socket_poll_prepare(struct modem_socket_config *cfg, struct modem_sock
 			      struct zsock_pollfd *pfd, struct k_poll_event **pev,
 			      struct k_poll_event *pev_end)
 {
+	bool event_handled = false;
```

The changes in this file don't seem right, it introduces inconsistency in poll() handling:
* `poll()` called with `POLLOUT` will give an error
* `poll()` called with `POLLIN | POLLOUT` will silently ignore `POLLOUT`

---

I had to include those changes to work around the net_lwm2m_engine: Error in poll:1 issue so the LwM2M sample would actually run. I don’t have a enough of experience with socket programming and zephyr modem drivers, so I leaned on the hl7800 driver, SIMCom SIM7080 and u-blox SARA-R4 drivers as references while building this one. Honestly not sure what a better solution would be at this point but if you’ve got any ideas or can point me in the right direction, I’m totally up for digging into it! Does anyone know if current drivers for sim7080 and ublox sara supports lwm2m?

```
[00:00:25.050,000] <dbg> hl78xx_socket: modem_pipe_callback: 359 modem_pipe_callback Pipe event received: 2
[00:00:25.062,000] <dbg> hl78xx_socket: modem_pipe_callback: 359 modem_pipe_callback Pipe event received: 1
[00:00:25.063,000] <dbg> hl78xx_socket: modem_process_handler: Processing handler - Bytes to read: 94, Collected: 0, Buffer size: 512
[00:00:25.063,000] <dbg> hl78xx_socket: modem_process_handler: 341 modem_process_handler 94 94 42
[00:00:25.101,000] <dbg> hl78xx_socket: modem_pipe_callback: 359 modem_pipe_callback Pipe event received: 1
[00:00:25.101,000] <dbg> hl78xx_socket: modem_process_handler: Processing handler - Bytes to read: 94, Collected: 42, Buffer size: 470
[00:00:25.101,000] <dbg> hl78xx_socket: modem_process_handler: 341 modem_process_handler 52 94 94
[00:00:25.101,000] <dbg> hl78xx_socket: on_cmd_sockread_common: Received Data:
                                        0d 0a 43 4f 4e 4e 45 43  54 0d 0a 14 fe fd 00 00 |..CONNEC T.......
                                        00 00 00 00 00 03 00 01  01 16 fe fd 00 01 00 00 |........ ........
                                        00 00 00 00 00 28 00 01  00 00 00 00 00 00 dc b9 |.....(.. ........
                                        8a a7 ea 2e 2a 71 dc f2  fc d6 57 af 7a 7b 6c 6f |....*q.. ..W.z{lo
                                        ba 53 38 1f 7b 32 83 18  b9 4e a1 31 d9 69 2d 2d |.S8.{2.. .N.1.i--
                                        45 4f 46 2d 2d 50 61 74  74 65 72 6e 2d 2d       |EOF--Pat tern--  
[00:00:25.101,000] <dbg> hl78xx_socket: on_cmd_sockread_common: Processed Data
                                        14 fe fd 00 00 00 00 00  00 00 03 00 01 01 16 fe |........ ........
                                        fd 00 01 00 00 00 00 00  00 00 28 00 01 00 00 00 |........ ..(.....
                                        00 00 00 dc b9 8a a7 ea  2e 2a 71 dc f2 fc d6 57 |........ .*q....W
                                        af 7a 7b 6c 6f ba 53 38  1f 7b 32 83 18 b9 4e a1 |.z{lo.S8 .{2...N.
                                        31 d9 69                                         |1.i              
[00:00:25.101,000] <dbg> hl78xx_socket: offload_recvfrom: 923 offload_recvfrom 0 67
z_impl_zsock_connect: 0 0
[00:00:25.102,000] <dbg> net_lwm2m_engine: lwm2m_socket_start: 1230 lwm2m_socket_start
[00:00:25.102,000] <dbg> hl78xx_socket: offload_ioctl: 1138 offload_ioctl 3
[00:00:25.102,000] <dbg> net_lwm2m_engine: lwm2m_socket_start: 1237 lwm2m_socket_start 0
[00:00:25.102,000] <dbg> hl78xx_socket: offload_ioctl: 1138 offload_ioctl 4
[00:00:25.102,000] <dbg> hl78xx_socket: offload_ioctl: F_SETFL called with flags=0x4000
[00:00:25.102,000] <dbg> net_lwm2m_engine: lwm2m_socket_start: 1244 lwm2m_socket_start 0
[00:00:25.102,000] <inf> net_lwm2m_engine: Connected, sock id 2
[00:00:25.102,000] <dbg> modem_chat: modem_chat_on_unknown_command_received: OK
[00:00:25.102,000] <dbg> modem_chat: modem_chat_log_received_command: +KUDP_RCV:  "23.97.187.154" 5684
[00:00:25.102,000] <dbg> hl78xx_dev: hl78xx_on_udprcv: 350 3 [+KUDP_RCV: ] ["23.97.187.154"] [5684]
[00:00:25.102,000] <dbg> net_lwm2m_registry: lwm2m_engine_get: path:1/0/1/0, level 3, buf:0x2001f7e0, buflen:4
[00:00:25.103,000] <dbg> net_lwm2m_rd_client: sm_send_registration: registration sent [23.97.187.154]
[00:00:25.103,000] <dbg> hl78xx_socket: offload_ioctl: 1138 offload_ioctl 258
[00:00:25.103,000] <dbg> hl78xx_socket: offload_ioctl: poll_prepare: fd=2, events=0x5
[00:00:25.103,000] <err> net_lwm2m_engine: Error in poll:1
[00:00:25.603,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 8
[00:00:25.603,000] <dbg> hl78xx_socket: offload_ioctl: 1138 offload_ioctl 258
[00:00:25.603,000] <dbg> hl78xx_socket: offload_ioctl: poll_prepare: fd=2, events=0x5
[00:00:25.603,000] <err> net_lwm2m_engine: Error in poll:1
[00:00:26.103,000] <dbg> hl78xx_socket: offload_ioctl: 1138 offload_ioctl 258
[00:00:26.103,000] <dbg> hl78xx_socket: offload_ioctl: poll_prepare: fd=2, events=0x5
[00:00:26.103,000] <err> net_lwm2m_engine: Error in poll:1
```

---

Pehraps the modem socket implementation could ignore `POLLOUT` and simply indicate that socket is always writeble, we actually do that in regular socket implementation for UDP:
https://github.com/zephyrproject-rtos/zephyr/blob/88b108d201c696667db123d571562f1c9f81f2ac/subsys/net/lib/sockets/sockets_inet.c#L1609
IMHO that would be a better approach than introducing such uncertainty in `poll()` behavior. But the decision is up to modem subsys maintainers.